src/main/java/com/yahoo/sketches/theta/ConcurrentUpdateSketch.java

```diff
@@ -0,0 +1,83 @@
+package com.yahoo.sketches.theta;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author eshcar
+ */
+public class ConcurrentUpdateSketch extends UpdateSketchComposition {
+
+  private static int PARALLELISM_LEVEL = 3;
+  protected static ExecutorService PROPAGATION_EXECUTOR_SERVICE =
+      Executors.newWorkStealingPool(PARALLELISM_LEVEL);
+
+  private volatile long theta_;
+  private volatile double estimation_;
+  // A flag to coordinate between several propagation threads
+  private AtomicBoolean propagationInProgress_;
+
+  // package visibility constructors - to be created only by factory
+  ConcurrentUpdateSketch(final UpdateSketch delegattee) {
+    super(delegattee);
+    theta_ = getThetaLong();
+    estimation_ = 0;
+    propagationInProgress_ = new AtomicBoolean(false);
+  }
+
+  public void propagate(final Sketch sketchIn, final AtomicBoolean propagationInProgress) {
+    final BackgroundThetaPropagation job = new BackgroundThetaPropagation(sketchIn,
+        propagationInProgress);
+    PROPAGATION_EXECUTOR_SERVICE.execute(job);
+  }
+
+  public double getEstimationSnapshot() {
+    return estimation_;
+  }
+
+  public long getVolatileTheta() {
+    return theta_;
+  }
+
+  private class BackgroundThetaPropagation implements Runnable {
+
+    private Sketch sketch_;
+    private AtomicBoolean localThreadPropagationFlag_;
+
+    public BackgroundThetaPropagation(final Sketch sketch,
+        final AtomicBoolean localThreadPropagationFlag) {
+      sketch_ = sketch;
+      localThreadPropagationFlag_ = localThreadPropagationFlag;
+    }
+
+    @Override public void run() {
+      assert getVolatileTheta() <= sketch_.getThetaLong();
+
+      while (!propagationInProgress_.compareAndSet(false,true)) {
+        //busy wait until can propagate
+      }
+      //At this point we are sure only a single thread is propagating data to the shared sketch
+
+      // propagate values from input sketch one by one
+      final long[] cacheIn = sketch_.getCache();
+      final int numEntries = sketch_.getRetainedEntries(false);
+      for (int i = 0; i < numEntries; i++) {
+        // pre-condition: cacheIn values are sorted in ascending order
+        final long hashIn = cacheIn[i];
+        if (hashIn >= getVolatileTheta()) {
```

Comment:
Since we know we are operating in a single thread here, why don't we keep the volatile theta in a local long and then store it back into the volatile at the end of the propagation?

Also, I don't see a number of checks that are performed in the UnionImpl merge code:

1) if the sketch is off-heap, you need to access the values via Memory.

2) The UnionImpl code performs the "Theta rule" before and after each sketch merge.  I don't see it here at all.  I don't think you can assume that the lowest theta is always the input sketch (or the shared sketch).

3) You are compacting each input sketch prior to your merge, (In order to leverage the early stop, I presume) But it is not at all clear that this is a good performance trade-off.  Have you characterized this?

***
Why are you updating the estimate after every merge ?

Reply:
I will check all the points you raise here .
Not sure what you mean by Theta rule, but we maintain an invariant  - the shared theta is always smaller or equal to the local theta

Reply:
Given any set operation on A, B: (union, intersection, difference), the "Theta rule" is 
result.theta = min( A.theta, B.theta ).  In UnionImpl line 239 and line 273, the Theta rule is applied both before and after the merge.  

What I see in your code is that you `assert getVolatileTheta() <= sketch_.getThetaLong();`, (which is not there at runtime) and at the end of the propagate you `theta_ = getThetaLong();`.   Why bother with the assert?  Just force the target theta to be the minimum of the two.  Also the `theta_ = getThetaLong();` is clearly one-sided.  It makes my head swim as to whether the Theta rule is actually being enforced.  The Theta rule IS the invariant that must be enforced, so have the code reflect that.

Reply:
I think I see another problem here, but perhaps it is because I don't completely understand what you are doing.

The Theta Union algorithm, which you appear to be emulating here, involves 3 theta values: the source, _thetaLongIn_, the destination, _gadget.getThetaLong()_, and _unionThetaLong_, which is derived from the other two.  When a new Union is created, the _unionThetaLong_ is always 1.0. 

Before the merge loop:  

*unionThetaLong_ = min(unionThetaLong_, thetaLongIn); //Theta rule*

After the merge loop:

*unionThetaLong_ = min(unionThetaLong_, gadget_.getThetaLong()); //sync thetaLongs*

Note that during the merge loop, the target gadget sketch always has integrity ... i.e., never has any dirty values .... A Theta-QuickSelect Sketch never has dirty values.  Only the Theta-Alpha sketch can have dirty values.   The Alpha sketch can never be used as a union target (gadget) only the QuickSelect sketch can.  

When the user request the result, it is always a Compact Sketch, because at this point the _unionThetaLong_ could be smaller than the gadget's thetaLong, which would invalidate (make dirty) some of the values held by the gadget.  So during the creation of the result Compact Sketch only the valid hash values are inserted into the Compact Sketch.

At no point in the processing are any of the sketches in an invalid state (with dirty values).

It is not at all clear to me how your are maintaining the integrity of the sketches during the concurrent processing.